### PR TITLE
[RTL] - Remove timing loop in LSU

### PR DESCRIPTION
**Pull request by tomroberts-lowrisc**
- External signals data_rvalid_i and data_err_i were factored
  into the external data_req_o signal
- To improve timing, these signals are decoupled
- The second part of an unaligned transaction will now be issued
  even if the first received an error response
- The state machine will service the abandoned requests
- pmp_err_q fixed to only update at specific times

### CODE REVIEW

File: rtl/ibex_load_store_unit.sv [hunk 1]
```diff
@@ -451,7 +503,8 @@ module ibex_load_store_unit (
   assign load_err_o    = data_or_pmp_err & ~data_we_q;
   assign store_err_o   = data_or_pmp_err &  data_we_q;
 
-  assign busy_o = (ls_fsm_cs == WAIT_RVALID) | (data_req_o == 1'b1);
+  assign busy_o = (ls_fsm_cs == WAIT_RVALID) | (ls_fsm_cs == WAIT_RVALID_ERR) |
+                  (ls_fsm_cs == WAIT_RVALID_DONE) | (data_req_o == 1'b1);
```
Comment (vogelpi) [inline rtl/ibex_load_store_unit.sv]: Should we just change this to `assign busy_o = (ls_fsm_cs != IDLE);`?

File: rtl/ibex_load_store_unit.sv [hunk 2]
```diff
@@ -397,6 +411,44 @@ module ibex_load_store_unit (
...
+        if (data_rvalid_i) begin
+          // Update the pmp error for the second part
+          pmp_err_d = data_pmp_err_i;
+          // The first part cannot see a PMP error in this state
+          if (data_err_i) begin
+            // first part created an error, abort transaction and wait for second rvalid
+            data_valid_o        = 1'b1;
+            data_or_pmp_err     = 1'b1;
+            handle_misaligned_d = 1'b0;
+            ls_fsm_ns           = WAIT_RVALID_ERR;
```
Comment (vogelpi) [inline rtl/ibex_load_store_unit.sv]: I think this will lock up the whole core: in case you get a bus error in the second part of a misaligned transaction (signaled together with `rvalid_i`) there will be no other `rvalid_i` from the bus. The FSM will stay inside `WAIT_RVALID_ERR` forever.

File: doc/load_store_unit.rst [hunk 3]
```diff
@@ -48,6 +48,10 @@ The LSU is able to handle misaligned memory accesses, meaning accesses that are
 However, it does so by performing two separate word-aligned accesses.
 This means that at least two cycles are needed for misaligned loads and stores.
 
+If an error response is received for the first transaction, the second transaction will still be issued.
+The second transaction will then follow the normal bus protocol, but it's response/data will be ignored.
```
Comment (imphil) [inline doc/load_store_unit.rst]: it's -> its

File: rtl/ibex_load_store_unit.sv [hunk 4]
```diff
@@ -370,7 +383,10 @@ module ibex_load_store_unit (
           end
         end else begin
           // first part rvalid is NOT received
-          ls_fsm_ns = WAIT_RVALID_MIS;
+          if (data_gnt_i) begin
+            // second grant is received
+            ls_fsm_ns = WAIT_RVALID_DONE;
+          end
         end
       end
```
Comment (vogelpi) [inline rtl/ibex_load_store_unit.sv]: The FSM was not very nice before I know, but I don't like that the super complex `WAIT_RVALID_MIS` state we have now. There are now three nested levels of if/else and so many statements, it is really hard to follow. You should at least add an empty line here and there to visually group what does (NOT) belong together. For example: what is below Line 357 does not depend on what is above. The same holds for Line 384.

File: doc/load_store_unit.rst (hunk 3 above)
Comment (tomroberts-lowrisc) [reply doc/load_store_unit.rst]: fixed

File: rtl/ibex_load_store_unit.sv (hunk 1 above)
Comment (tomroberts-lowrisc) [reply rtl/ibex_load_store_unit.sv]: agreed

File: rtl/ibex_load_store_unit.sv (hunk 4 above)
Comment (tomroberts-lowrisc) [reply rtl/ibex_load_store_unit.sv]: You're right, treating the two errors differently was over-complicating things. I have made the behaviour the same and simplified the WAIT_RVALID_MIS state a bit.

File: rtl/ibex_load_store_unit.sv (hunk 2 above)
Comment (tomroberts-lowrisc) [reply rtl/ibex_load_store_unit.sv]: why wouldn't there be a second rvalid_i?

Comment (vogelpi) [reply rtl/ibex_load_store_unit.sv]: You are right. I thought the first `rvalid_i` would have been received already when jumping into `WAIT_RVALID_DONE`. Sorry!

File: rtl/ibex_load_store_unit.sv (hunk 4 above)
Comment (vogelpi) [reply rtl/ibex_load_store_unit.sv]: Well done, it looks much better now!

File: rtl/ibex_load_store_unit.sv [hunk 5]
```diff
@@ -345,32 +349,34 @@ module ibex_load_store_unit (
...
 
       WAIT_RVALID_MIS: begin
+        // push out second request, unless the first request was suppressed by a PMP error
+        data_req_o = 1'b1;
         // tell ID/EX stage to update the address
         addr_incr_req_o = 1'b1;
+
         // first part rvalid is received, or gets a pmp error
```
Comment (vogelpi) [inline rtl/ibex_load_store_unit.sv]: pmp -> PMP

File: rtl/ibex_load_store_unit.sv [hunk 6]
```diff
@@ -397,6 +401,44 @@ module ibex_load_store_unit (
         end
       end
 
+      WAIT_GNT_ERR: begin
+        // Wait for the grant of the abandoned second access
+        data_req_o = 1'b1;
+        // tell ID/EX stage to update the address
+        addr_incr_req_o = 1'b1;
+        if (pmp_err_q) begin
+          // The second part was suppressed by a pmp error
```
Comment (vogelpi) [inline rtl/ibex_load_store_unit.sv]: pmp -> PMP

File: rtl/ibex_load_store_unit.sv [hunk 7]
```diff
@@ -345,32 +349,34 @@ module ibex_load_store_unit (
...
+        data_req_o = 1'b1;
         // tell ID/EX stage to update the address
         addr_incr_req_o = 1'b1;
+
         // first part rvalid is received, or gets a pmp error
-        // pmp_err_i will hold stable until the address is updated, and
-        // therefore pmp_err_q is valid in both WAIT_GNT_MIS and WAIT_RVALID_MIS states
         if (data_rvalid_i || pmp_err_q) begin
+          // Update the PMP error for the second part
+          pmp_err_d = data_pmp_err_i;
```
Comment (vogelpi) [inline rtl/ibex_load_store_unit.sv]: Question: shouldn't we always update `pmp_err_d` with `data_pmp_err_i` in this state, in particular also when we do not get the `rvalid_i`, i.e., when we go to line 374? As far as I understand, we do a new PMP check in this state and would like to buffer the outcome of that check.

Comment (tomroberts-lowrisc) [reply rtl/ibex_load_store_unit.sv]: No we can't update it yet - if the first request did not get a PMP error (and is just sitting there waiting for rvalid) but the second request will get a PMP error, if we update the pmp_err_q flop here, then we will see pmp_err_q set on the following cycle and assume it relates to the first request and take the error without waiting for the first rvalid to come back.

Comment (vogelpi) [reply rtl/ibex_load_store_unit.sv]: I see, thanks for clarifying this to me.

**Comment by arjanbink** [comment]
Thank you for the proposed update. I don't see anything wrong with it (and it would remove the combinational path on the data interface that I asked about; but of course not yet the similar path on the instruction side). 

It would help me however if you could describe the LSU protocol with respect to data_pmp_err_i signal (I see pmp_err_q being used together with data_err_i, data_gnt_i, as well as data_rvalid_i).

**Comment by tomroberts-lowrisc** [comment]
> Thank you for the proposed update. I don't see anything wrong with it (and it would remove the combinational path on the data interface that I asked about; but of course not yet the similar path on the instruction side).
> 
> It would help me however if you could describe the LSU protocol with respect to data_pmp_err_i signal (I see pmp_err_q being used together with data_err_i, data_gnt_i, as well as data_rvalid_i).

I'll add some extra comments to describe the timing / functionality of the two types of error

**Comment by vogelpi** [comment]
Thanks @arjanbink for your feedback and @tomroberts-lowrisc for adding the clarifying comment.

Shall we finally merge this PR?

**Comment by arjanbink** [comment]
Okay with me; thank you for the quick follow-up
